File: 2015/solutionsR/day02.R
# Day 2 - Part 1 ----------------------------------------------------------

input <- readLines("2015/data/day02.txt", warn = F) # load the data
input <- as.data.frame(input)                       # convert to df 1000x1
input <- strsplit(input[[1]], "x")                  # split by "x" to get each side separately
input <- lapply(input, as.integer)                  # convert to integers
input <- do.call(rbind, input)                      # 1000x3 df
names(input) <- c("w", "l", "h")                    # column names - sides - width, length, height

lowest <- apply(input, 1, function(x) (sort(x))[1]) # find the lowest number (smallest side)
lowest2<- apply(input, 1, function(x) (sort(x))[2]) # find the 2nd lowest number

input$extra <- lowest * lowest2                     # extra paper is the product/area of the smallest sides

input$paper <- (2*input$w*input$l)+(2*input$l*input$h)+(2*input$w*input$h)+input$extra # 2*l*w + 2*w*h + 2*h*l + extra

sum(input$paper)        

# Day 2 - Part 2 ----------------------------------------------------------

input <- readLines("2015/data/day02.txt", warn = F) # load the data
input <- as.data.frame(input)                       # convert to df 1000x1
input <- strsplit(input[[1]], "x")                  # split by "x" to get each side separately
input <- lapply(input, as.integer)                  # convert to integers
input <- do.call(rbind, input)                      # 1000x3 df
names(input) <- c("w", "l", "h")                    # column names - sides - width, length, height

lowest <- apply(input, 1, function(x) (sort(x))[1]) # find the lowest number (smallest side)
lowest2<- apply(input, 1, function(x) (sort(x))[2]) # find the 2nd lowest number

input$wrap <- (lowest*2)+(lowest2*2)                # shortest distance around (smallest) sides
input$bow <- input$w*input$l*input$h                # volume

input$ribbon <- input$wrap + input$bow              # put together to get the ribbon length needed

sum(input$ribbon)

File: 2015/solutionsPython/day02.py
# Day 2 - Part 1 ----------------------------------------------------------
with open("2015/data/day02.txt") as data:
    aoc_input = data.read().splitlines()

# done through list comprehension

aoc_input = [i.split("x") for i in aoc_input]                         # split into three-element-sublists
aoc_input = [sorted([int(j) for j in i]) for i in aoc_input]          # sort dimensions in each sublist

sides = [2*e[0]*e[1] + 2*e[1]*e[2] + 2*e[2]*e[0] for e in aoc_input]  # sublist == all sides combined
extra = [elem[0] * elem[1] for elem in aoc_input]                     # extra paper is the product of 2 shortest sides
wrapper = [sides[i] + extra[i] for i in range(len(sides))]            # sum sides and extra paper of each present

sum(wrapper)                                                          # sum the list to get all paper needed

# Day 2 - Part 2 ----------------------------------------------------------
with open("2015/data/day02.txt") as data:
    aoc_input = data.read().splitlines()

aoc_input = [i.split("x") for i in aoc_input]                         # split into three-element-sublists
aoc_input = [sorted([int(j) for j in i]) for i in aoc_input]          # sort dimensions in each sublist

dist = [elem[0]*2 + elem[1]*2 for elem in aoc_input]                  # 2 shortest distances across smallest sides
bow = [elem[0]*elem[1]*elem[2] for elem in aoc_input]                 # bow of each present as defined

ribbon = [dist[i] + bow[i] for i in range(len(dims))]                 # sum the distances and bow of each present

sum(ribbon)                                                           # sum the list to get all ribbon needed
